consensus/hotstuff/safetyrules/safety_rules_test.go

```diff
@@ -1,165 +1,526 @@
+// TestProduceVote_VotingOnUnsafeProposal tests different scenarios where we try to vote on unsafe blocks
+func (s *SafetyRulesTestSuite) TestProduceVote_VotingOnUnsafeProposal() {
+	s.Run("invalid-block-view", func() {
+		// create block with block.View == block.QC.View
+		proposal := helper.MakeProposal(
+			helper.WithBlock(
+				helper.MakeBlock(
+					helper.WithParentBlock(s.bootstrapBlock),
+					helper.WithBlockView(s.bootstrapBlock.View))))
+		vote, err := s.safety.ProduceVote(proposal, proposal.Block.View)
+		require.Error(s.T(), err)
+		require.Nil(s.T(), vote)
+	})
+	s.Run("view-already-acknowledged", func() {
+		// create block with view <= HighestAcknowledgedView
+		proposal := helper.MakeProposal(
+			helper.WithBlock(
+				helper.MakeBlock(
+					func(block *model.Block) {
+						block.QC = helper.MakeQC(helper.WithQCView(s.safetyData.HighestAcknowledgedView - 1))
+					},
+					helper.WithBlockView(s.safetyData.HighestAcknowledgedView))))
+		vote, err := s.safety.ProduceVote(proposal, proposal.Block.View)
+		require.True(s.T(), model.IsNoVoteError(err))
+		require.Nil(s.T(), vote)
+	})
+	s.Run("no-last-view-tc", func() {
+		// create block where Block.View != Block.QC.View+1 and LastViewTC = nil
+		proposal := helper.MakeProposal(
+			helper.WithBlock(
+				helper.MakeBlock(
+					helper.WithParentBlock(s.bootstrapBlock),
+					helper.WithBlockView(s.bootstrapBlock.View+2))))
+		vote, err := s.safety.ProduceVote(proposal, proposal.Block.View)
+		require.Error(s.T(), err)
+		require.Nil(s.T(), vote)
+	})
+	s.Run("last-view-tc-invalid-view", func() {
+		// create block where Block.View != Block.QC.View+1 and
+		// Block.View != LastViewTC.View+1
+		proposal := helper.MakeProposal(
+			helper.WithBlock(
+				helper.MakeBlock(
+					helper.WithParentBlock(s.bootstrapBlock),
+					helper.WithBlockView(s.bootstrapBlock.View+2))),
+			helper.WithLastViewTC(
+				helper.MakeTC(
+					helper.WithTCView(s.bootstrapBlock.View))))
+		vote, err := s.safety.ProduceVote(proposal, proposal.Block.View)
+		require.Error(s.T(), err)
+		require.Nil(s.T(), vote)
+	})
+	s.Run("last-view-tc-invalid-highest-qc", func() {
+		// create block where Block.View != Block.QC.View+1 and
+		// Block.View == LastViewTC.View+1 and Block.QC.View < LastViewTC.NewestQC.View
+		// in this case block is not safe to extend since proposal is built on top of QC, which is lower
+		// than QC presented in LastViewTC.
+		TONewestQC := helper.MakeQC(helper.WithQCView(s.bootstrapBlock.View + 1))
+		proposal := helper.MakeProposal(
+			helper.WithBlock(
+				helper.MakeBlock(
+					helper.WithParentBlock(s.bootstrapBlock),
+					helper.WithBlockView(s.bootstrapBlock.View+2))),
+			helper.WithLastViewTC(
+				helper.MakeTC(
+					helper.WithTCView(s.bootstrapBlock.View+1),
+					helper.WithTCHighestQC(TONewestQC))))
+		vote, err := s.safety.ProduceVote(proposal, proposal.Block.View)
+		require.Error(s.T(), err)
```

Comment:
Exceptions should not be a `NoVoteError`:
```suggestion
		require.Error(s.T(), err)
		require.False(s.T(), model.IsNoVoteError(err))
```